runtime/dict.go

```diff
@@ -624,6 +624,34 @@ func dictKeys(f *Frame, args Args, kwargs KWArgs) (*Object, *BaseException) {
 	return toDictUnsafe(args[0]).Keys(f).ToObject(), nil
 }
 
+func dictPop(f *Frame, args Args, _ KWArgs) (*Object, *BaseException) {
+	expectedTypes := []*Type{DictType, ObjectType, ObjectType}
+	argc := len(args)
+	if argc == 2 {
+		expectedTypes = expectedTypes[:2]
+	}
+	if raised := checkMethodArgs(f, "pop", args, expectedTypes...); raised != nil {
+		return nil, raised
+	}
+	key := args[1]
+	d := toDictUnsafe(args[0])
+	d.mutex.Lock(f)
+	item, raised := d.GetItem(f, key)
```

Comment:
As you mention, calling GetItem then DelItem requires multiple lookups which includes hashing, equality check, etc. It'd be better to call lookupEntry() directly so that you can then remove it without having to look it up again.

One possible approach would be to modify putItem() so that it works when passed value == nil. Then, rather than returning a bool, have it return the original entry that was present (or nil if entry == deletedEntry). That would tell the caller whether something was overwritten or not which is why it returns a bool now.

Then DelItem and PopItem could both call putItem.